Add local file option and formatter

AlecThomson opened the pull request:
Allows local file support (i.e. if you `wget` a bunch of IONEX data for offline use).

Adds a formatter for the ever-changing IONEX formats.

Also some tweaks to typing for dev use

Closes #46

maaijke commented on RMextract/getIONEX.py:
```diff
@@ -680,51 +714,80 @@ def get_urllib_IONEXfile(time="2012/03/23/02:20:10.01",
-        else:
-            os.remove(fname)
-        fname=fname[:-2]
+    out_fname = fname[:-2] if fname[-2:].upper() == '.Z' else fname    
```
I think the new IONEX filename format uses .gz instead of .Z, maybe we can already anticipate that here?

maaijke commented on RMextract/getIONEX.py:
```diff
@@ -680,51 +714,80 @@ def get_urllib_IONEXfile(time="2012/03/23/02:20:10.01",
+    out_fname = fname[:-2] if fname[-2:].upper() == '.Z' else fname    
+    
+    # First, if the final file already exists, simply return
+    if os.path.exists(out_fname):
```
only if overwrite is not True?

AlecThomson commented:
@maaijke - I've reworked a few things, which should address your comments. I've also added some more standard use of `pathlib` and `urllib`. Although, I haven't been able to do this throughout the entire codebase. The former makes life much easier for dealing with paths (over using strings and `os.path`).

I've added a new `formatters` submodule, which should lay things out more explicitly for a user. I've also dropped the redundant `yy` from the format.

On the change of filename extension, I think it should be easy enough to do in the new submodule. For example (that I'm making up), if the extensions change after 2025 we could add something like:
```python
def ftp_aiub_unibe_ch(server: str, year: int, dayofyear: int, prefix: str) -> str:
    """
    Format the URL for the ftp.aiub.unibe.ch server
    """
    ext = ".Z" if year <= 2024 else ".gz"
    return f"{server}/CODE/{year:4d}/{prefix.upper()}{dayofyear:03d}0.{year%100:02d}I{ext}"
```

AlecThomson commented:
Oh, I also did a quick bit of autoformatting to make my IDE happy. Sorry if this touched a few files. This mostly fixed up the order of imports (compliant with `isort`), so shouldn't change any actual code behaviour.

gmloose reviewed (commented):
You made a lot of good improvements. This is very much appreciated, because the code needed some extra love and care. I made a few minor remarks.

I saw in the email conversation you also suggested to use a linting tool like `ruff`, and to use more rigorous code formatting (I would suggest to use `black` for that). I can only support that. 
Also, I think the package desperately needs tests. I consider that even more important than proper formatting.

Please feel free to become more closely involved in improving `RMextract`.

gmloose commented:
One thing that bothers me a bit, but is unrelated to your changes is the dependency on `PySocks`. That package seems to be unmaintained. The latest release was made in 2019, and I read somewhere that people already run into issues with Python 3.10. I think it should be replaced with something more modern (don't know yet which package that would be), or we should fall back to the `socket` package from the Python standard library.

AlecThomson commented:
> You made a lot of good improvements. This is very much appreciated, because the code needed some extra love and care. I made a few minor remarks.
> 
> I saw in the email conversation you also suggested to use a linting tool like `ruff`, and to use more rigorous code formatting (I would suggest to use `black` for that). I can only support that. Also, I think the package desperately needs tests. I consider that even more important than proper formatting.
> 
> Please feel free to become more closely involved in improving `RMextract`.

Thanks @gmloose! Happy to help out where I can!

A few top-level thoughts and some additional context. My changes here started from using RMextract in a) a highly parallel manner, and b) an environment where FTP was blocked on compute nodes. This is what lead me to adding the formatting option (allowing for easier offline file access) and catching some race conditions. I'll admit that perhaps this PR has grown a little larger than that narrower set of objectives. For the most part, I tried not to touch any of the actual logic in the existing code.

@maaijke mentioned that some more major reworks are potentially in order as well, which added some more motivation for me not to change much of the logical code - rather I tried to make the existing logic clearer for any devs / contributors (starting with myself!). I'm happy to provide some suggestions on how a rework might look - but I can't commit the time to the actual implementation. Plus, I think there'd need to be clear agreement from the core maintainers about how to go about that.

I'm happy to start adding tests for the functions I've 'touched' along the way. I've found this to be the easiest way to get unit tests in for a large project that has grown without them. Some less unitary end-to-end tests are useful as well. Perhaps after this PR merge we can start looking at creating those frameworks.

On formatters, I'm also a fan of Black. However, I've recently made the move to Ruff as it seems to be where the wider Python community is moving - and the completely replicates the feature set with greater speed.

> One thing that bothers me a bit, but is unrelated to your changes is the dependency on PySocks. That package seems to be unmaintained. The latest release was made in 2019, and I read somewhere that people already run into issues with Python 3.10. I think it should be replaced with something more modern (don't know yet which package that would be), or we should fall back to the socket package from the Python standard library.

I added `socks` to the `setup.py` as it was already a silent dependency in some of the functions. However, I think it went unnoticed since it was buried in proxy server logic.

AlecThomson commented on RMextract/getIONEX.py:
```diff
@@ -626,105 +613,120 @@ def get_urllib_IONEXfile(time="2012/03/23/02:20:10.01",
+        #try primary url
+        try:
+            _ = request.urlopen(server,timeout=30)
+        except Exception as e:
```
This is a bare exception in the current branch - see: https://github.com/lofar-astron/RMextract/blob/25638b1f8544083c388e0c699cfeb503ba7b8147/RMextract/getIONEX.py#L690
                
I'm not sure what errors will be raised by a non-response, which is why I added the logging of the error. Perhaps we could simply re-raise after the 'backup' fails?

AlecThomson commented on RMextract/PosTools.py:
```diff
@@ -1,42 +1,49 @@
-from scipy import interpolate
+    import ephem
+    if not HAS_PYRAP:
+        logger.debug('PyEphem will be used to perform calculations!')
```
I put these at debug level since they occur on import. Previously, for any dependent module (e.g. [FRion](https://github.com/CIRADA-Tools/FRion)) or a multiprocessing invocation these messages will appear _a lot_.

AlecThomson commented on RMextract/getIONEX.py:
```diff
@@ -626,105 +613,120 @@ def get_urllib_IONEXfile(time="2012/03/23/02:20:10.01",
+    if not overwrite and out_fname.exists():
+        return out_fname.as_posix()
+
+    # Here we actually download the file. Lets make sure though that this
+    # file is not downloaded by another task first
```
Indeed this was because we hit race conditions during parallel execution. Although, I think this use-case is possibly unusual. Our current workaround is to execute a singular task, which initialises the files, then trigger the full parallel execution which is fine with the existence check.

AlecThomson commented on RMextract/getRM.py:
```diff
@@ -1,144 +1,161 @@
+    use_mean: Optional[bool] = None,
+    stat_names=[],
+    useEMM=False,
+    object = '',
```
Agreed - this argument was copied from the existing hidden kwarg:
https://github.com/lofar-astron/RMextract/blob/25638b1f8544083c388e0c699cfeb503ba7b8147/RMextract/getRM.py#L59

Happy to change, but it could break existing use elsewhere.a

AlecThomson commented on RMextract/getIONEX.py:
```diff
@@ -626,105 +613,120 @@ def get_urllib_IONEXfile(time="2012/03/23/02:20:10.01",
+                    server=backupserver
+                    logger.warning(f"Primary IONEX host '{server}' resolution "
+                                                f"failure. Trying backup at '{backupserver}'")
+                except Exception as e:
```
As above

AlecThomson commented on RMextract/getIONEX.py:
```diff
@@ -626,105 +613,120 @@ def get_urllib_IONEXfile(time="2012/03/23/02:20:10.01",
+    
+    url = formatter(server=server, prefix=prefix, year=year, dayofyear=dayofyear)
+
+    logger.info(f"Constructed {url=}.")
```
As above

AlecThomson commented on RMextract/PosTools.py:
```diff
@@ -1,42 +1,49 @@
+    import casacore.quanta as qu
+    from casacore import tables as tab
+    from casacore.measures import measures
+    logger.debug('casacore will be used to compute positions')
+except ImportError:
+    logger.debug('We will need PyEphem to perform calculations!')
+    logger.debug('the accuracy of results might decease a bit')
```
This is in the existing branch, but happy to fix
https://github.com/lofar-astron/RMextract/blob/25638b1f8544083c388e0c699cfeb503ba7b8147/RMextract/PosTools.py#L9

gmloose commented:
> One thing that bothers me a bit, but is unrelated to your changes is the dependency on `PySocks`. That package seems to be unmaintained. The latest release was made in 2019, and I read somewhere that people already run into issues with Python 3.10. I think it should be replaced with something more modern (don't know yet which package that would be), or we should fall back to the `socket` package from the Python standard library.

Funny. The well-known and often-used `requests` package added support for socks in version 2.10. And guess what..., for this it depends on the third-party package `PySocks`. 🤔

AlecThomson commented on RMextract/getRM.py:
```diff
@@ -1,144 +1,161 @@
+        logger.info("Using radec instead of pointing")
+        pointing = radec    
+
+    if start_time and end_time:
+        time_in_sec = False
+    elif (start_time and not end_time) or (end_time and not start_time):
+        raise ValueError("start_time and end_time must be given together")
```
Happy to add, but again this is verbatim copied from the original code

gmloose commented:
@maaijke,  do you want to press the  Merge button? Please do not forgot to select "Squash and merge".
